### ISSUE TITLE
Can protobuf.js be used together with grpc-web?

### ISSUE THREAD

**Issue by gunters63** [protobufjs/protobuf.js#1567]
I am currently using [grpc-web](https://github.com/grpc/grpc-web) in a React SPA-project.

The backend is implemented with [gRPC for .NET](https://github.com/grpc/grpc-dotnet), which already implements a middleware to translate gRPC-Web to gRPC HTTP/2 (like a built-in Envoy).

On the client side I use `protoc.exe` together with the grpc-web plugin (`protoc-gen-grpc-web.exe`) generate code for my services and for the proto serialization.

`protoc.exe` is currently not able to generate ES6 modules which is a bit problematic in my setup (I use [Vite](https://vitejs.dev/)). Furthermore protobufjs claims to be several times faster with serialization/deserialization than the `protoc.exe` generated code.

So my question: Is it possible to use protobuf.js together with grpc-web?

**Comment by alexander-fenster** [comment]
I _think_ it's not possible, but it's better to ask [grpc-web](https://github.com/grpc/grpc-web) folks :)

**Comment by gunters63** [comment]
I solved my problem in the meanwhile by hacking protoc.exe to generate ES6 modules

**Comment by achingbrain** [comment]
Not really as this module doesn't support streaming rpc methods properly - there's a PR to fix that here but it appears to have stalled: https://github.com/protobufjs/protobuf.js/pull/1115

**Comment by lijiaqigreat** [comment]
This might help: https://github.com/grpc/grpc-web/issues/80#issuecomment-426517834
It implements `rpcImpl` for grpc web, which is a little hacky and only handles non streaming.

**Comment by arontsang** [comment]
> I solved my problem in the meanwhile by hacking protoc.exe to generate ES6 modules

You might want to take a look at my nuget package https://www.nuget.org/packages/ArtsTech.Grpc.TypescriptClientGenerator/

**Comment by phi0411141** [comment]
@gunters63 can you help to share how to make protoc.exe generates es6 module?

**Comment by dhlolo** [comment]
> @gunters63 can you help to share how to make protoc.exe generates es6 module?

`npm i protobufjs -g & pbjs -t static-module -w es6 -o proto.js titan.proto --no-convert --no-delimited --no-comments`

**Comment by gunters63** [comment]
I have forked commit a1d8f8f9 and patched ES6 support. It works for me and sorry, I never bothered to pull master again.

Here is a patch:

https://gist.github.com/gunters63/3c8798de46fa469e1ce06966df017ca8

**Comment by gunters63** [comment]
Here a link: https://nextcloud.spranz.info/s/A7xHwmrbM6ByTnf

**Comment by nathansegers** [comment]
@gunters63 Thank you very much for this patch! This solved my search for a working Typescript version of my Stubs!

**Comment by gunters63** [comment]
@nathansegers: Yeah, looks like the Linux binaries I uploaded where built from unpatched sources. 

I really don't understand why the protobuf team didn't implement ES6 support yet. It should be quite easy.. There is some initial support (header constants ...) in the sources already for years now.

**Comment by nathansegers** [comment]
Currently I'll just leave to work on Windows, which is fine for my situation ;-)

**Comment by yinzara** [comment]
This is an interesting alternative now:
https://buf.build/blog/protobuf-es-the-protocol-buffers-typescript-javascript-runtime-we-all-deserve

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
